refactoring of the SYCL backend

AuroraPerego opened the pull request:
This refactoring aims to reduce code duplication and ease the process of adding new SYCL backends (e.g. `GpuSyclNvidia` and `GpuSyclAmd`).

main changes:
- removed the inheritance for the SYCL `Acc`, changed to `using`
- templated everything on the `Tag`, device selectors included, and moved some common methods / structs in the `*GenericSycl.hpp` files

AuroraPerego committed:
move the SYCL accelerators from inheritance to typedef

AuroraPerego committed:
remove code duplication in getFunctionAttributes for SYCL

AuroraPerego committed:
template SYCL objects and device selectors on the Tag

- template the Device, Queue, Event, Acc, Buf, Platform, EventHostManualTriggerSycl and device selectors on the Tag
- adapt the default queue for the tests
- avoid code duplication for SYCL where possible

psychocoderHPC commented:
@AuroraPerego could you please perform locally `git commit --amend  -a` and force push again to re-trigger the CI.
Simple restarting does not work, I am not sure if some caches avoid a clean re-test.

fwyzard commented:
At some point, could you also squash the last two commits into the third one ?

fwyzard commented on include/alpaka/acc/AccCpuSycl.hpp:
@AuroraPerego I think you should update the copyright notices on this and the other files you modified.

fwyzard commented on include/alpaka/acc/AccCpuSycl.hpp:
```diff
@@ -6,16 +6,7 @@
-#include "alpaka/platform/Traits.hpp"
-#include "alpaka/vec/Vec.hpp"
 
 #include <string>
 #include <utility>
```
I think these two headers can be removed (also in the other accelerator specialisations).
```suggestion
```

fwyzard commented on include/alpaka/platform/PlatformCpuSycl.hpp:
```diff
@@ -18,27 +18,20 @@ namespace alpaka
                 return dev.is_cpu() ? 1 : -1;
             }
+
+            static constexpr char name[] = "CpuSycl";
```
If this is only used in the accelerator specialisations, should it be moved there ?
Or, if it is used anywhere else, maybe it should be moved to the platform classes ?

AuroraPerego replied:
It's used only [here](https://github.com/AuroraPerego/alpaka/blob/SYCLrefactoring/include/alpaka/acc/AccGenericSycl.hpp#L167), so I can try to move it.

fwyzard commented on include/alpaka/mem/buf/BufGenericSycl.hpp:
```diff
@@ -197,38 +195,40 @@ namespace alpaka::trait
 
-            // Allocate SYCL page-locked memory on the host, mapped into the TPlatform address space and
-            // accessible to all devices in the TPlatform.
+            // Allocate SYCL page-locked memory on the host, mapped into the PlatformGenericSycl address space and
+            // accessible to all devices in the PlatformGenericSycl.
```
I would write more generically "platform" instead of "PlatformGenericSycl":
```suggestion
            // Allocate SYCL page-locked memory on the host, mapped into the SYCL platform's address space and
            // accessible to all devices in the SYCL platform.
```

fwyzard reviewed (changes requested):
See the other comments.

AuroraPerego committed:
address PR comments

- update copyright notices
- remove unnecessary includes
- take Acc name from the Tag
- improve comments

fwyzard commented on include/alpaka/acc/AccGenericSycl.hpp:
```diff
@@ -164,8 +164,8 @@ namespace alpaka::trait
         {
-            return std::string("Acc") + detail::SYCLDeviceSelector<TTag>::name + "<" + std::to_string(TDim::value)
-                   + "," + core::demangled<TIdx> + ">";
+            return std::string("Acc") + core::demangled<TTag>.substr(11) + "<" + std::to_string(TDim::value) + ","
```
😭

fwyzard replied:
```suggestion
            return std::string("Acc") + core::demangled<TTag>.substr(__builtin_strlen("alpaka::Tag")) + "<" + std::to_string(TDim::value) + ","
```
at least is self-documenting :-p

fwyzard reviewed (approved):
Note to whoever merges this PR: the commits should be squashed into a single one.
